src/backend/distributed/transaction/backend_data.c

```diff
@@ -878,55 +878,52 @@ AssignDistributedTransactionId(void)
 
 
 /*
- * AssignGlobalPID assigns a global process id for the current backend.
- * If this is a Citus initiated backend, which means it is distributed part of a distributed
- * query, then this function assigns the global pid extracted from the application name.
- * If not, this function assigns a new generated global pid.
+ * AssignGlobalPID assigns a global process id for the current backend based on
+ * the given applicationName. If this is a Citus initiated backend, which means
+ * it is distributed part of a distributed query, then this function assigns
+ * the global pid extracted from the application name. If not, this function
+ * assigns a new generated global pid.
  *
- * If a global PID is already assigned to this backend, then this function is a
- * no-op. In most scenarios this would already be the case, because a newly
- * assigned global PID would be the same as a proviously assigned one. But
- * there's two important cases where the newly assigned  global PID would be
- * different from the previous one:
- * 1. The current backend is an internal backend and in the meantime the
- *    application_name was changed to one without a gpid, e.g.
- *    citus_rebalancer. In this case we don't want to throw away the original
- *    gpid of the query originator, because that would mess up distributed
- *    deadlock detection involving this backend.
- * 2. The current backend is an external backend and the node id of the current
- *    node changed. Updating the gpid to match the nodeid might actually seem
- *    like a desirable property, but that's not the case. Updating the gpid
- *    with the new nodeid would mess up distributed deadlock and originator
- *    detection of queries too. Because if this backend already opened
- *    connections to other nodes, then those backends will still have the old
- *    gpid.
+ * There's one special case where we don't want to assign a new pid and keep
+ * the old pid on purpose: The current backend is an external backend and the
+ * node id of the current node changed. Updating the gpid to match the nodeid
+ * might seem like a desirable property, but that's not the case. Updating the
```

Comment:
We don't check the nodeid in this function. How do we know if the node id changed or not?

Reply:
So `GenerateGlobalPID` (which is called in this function) will internally call `GetLocalNodeId`. And the problem is that the return value of `GetLocalNodeId` can differ due to the current node being added on a subsequent call. This is a problem if this backend, because any open connection that this backend made before that will have the old gpid/nodeid in their application name. 

The way this function handles this case is not by reading the local nodeid, but simply checking if this backend already was a command originator and was assigned a gpid. Because in that case we don't want to change the gpid.